File: cpp/Linked_List/0707_Design_Linked_List.cpp

//leetcode submit region begin(Prohibit modification and deletion)
class MyLinkedList {
public:
  struct LinkedNode {
    int val;
    LinkedNode *next;

    LinkedNode(int val) : val(val), next(nullptr) {}
  };

  MyLinkedList() {
    m_dummyHead = new LinkedNode(0);
    m_size = 0;
  }

  int get(int index) {
    if (index < 0 || index > m_size - 1) return -1;
    LinkedNode *cur = m_dummyHead->next;
    while (index--) {
      cur = cur->next;
    }
    return cur->val;
  }

  void addAtHead(int val) {
    LinkedNode *newNode = new LinkedNode(val);
    newNode->next = m_dummyHead->next;
    m_dummyHead->next = newNode;
    m_size++;
  }

  void addAtTail(int val) {
    LinkedNode *newNode = new LinkedNode(val);
    LinkedNode *cur = m_dummyHead;
    while (cur->next) {
      cur = cur->next;
    }
    cur->next = newNode;
    m_size++;
  }

  void addAtIndex(int index, int val) {
    if (index > m_size || index < 0) return;
    LinkedNode *newNode = new LinkedNode(val);
    LinkedNode *cur = m_dummyHead;
    while (index--) {
      cur = cur->next;
    }
    newNode->next = cur->next;
    cur->next = newNode;
    m_size++;
  }

  void deleteAtIndex(int index) {
    if (index > m_size - 1 || index < 0) return;
    LinkedNode *cur = m_dummyHead;
    while (index--) {
      cur = cur->next;
    }
    LinkedNode *tmp = cur->next;
    cur->next = cur->next->next;
    delete tmp;
    m_size--;
  }

private:
  LinkedNode *m_dummyHead;
  int m_size;
};

/**
 * Your MyLinkedList object will be instantiated and called as such:
 * MyLinkedList* obj = new MyLinkedList();
 * int param_1 = obj->get(index);
 * obj->addAtHead(val);
 * obj->addAtTail(val);
 * obj->addAtIndex(index,val);
 * obj->deleteAtIndex(index);
 */
//leetcode submit region end(Prohibit modification and deletion)

File: python3/0707-design-linked-list.py
class Node:
    def __init__(self, val):
        self.val = val
        self.next = None


class MyLinkedList:

    def __init__(self):
        self._head = Node(0)
        self._count = 0

    def get(self, index: int) -> int:
        if 0 <= index < self._count:
            node = self._head
            for _ in range(0, index + 1):
                node = node.next
                index -= 1
            return node.val
        return -1

    def addAtHead(self, val: int) -> None:
        self.addAtIndex(0, val)

    def addAtTail(self, val: int) -> None:
        self.addAtIndex(self._count, val)

    def addAtIndex(self, index: int, val: int) -> None:
        if index > self._count:
            return
        self._count += 1
        node = Node(val)
        pre, cur = Node(0), self._head
        for _ in range(0, index + 1):
            pre = cur
            cur = cur.next
        pre.next = node
        node.next = cur

    def deleteAtIndex(self, index: int) -> None:
        if 0 <= index < self._count:
            self._count -= 1
            pre, cur = Node(0), self._head
            for _ in range(0, index + 1):
                pre = cur
                cur = cur.next
            pre.next = cur.next
            cur.next = None

# Your MyLinkedList object will be instantiated and called as such:
# obj = MyLinkedList()
# param_1 = obj.get(index)
# obj.addAtHead(val)
# obj.addAtTail(val)
# obj.addAtIndex(index,val)
# obj.deleteAtIndex(index)
